Introduce builddef.VersionMap struct

NiR- opened the pull request:
This new struct is used in specialized definitions whenever a map of packages/extensions/etc... has to be associated with version constraints. This struct has an `Add()` method which is used to define values without overwriting manually-defined ones. This was originally motivated by a bug in the php handler but this struct seems useful in other places.

Here's the original bug description:

> The infered extensions were overwriting manually defined version
constraints. Because of that, PHP prod stages were always resolving the
latest version of APCu, even when a specific version was pinned in the
zbuildfile.
>
> This new struct is used to properly add infered extensions without
overwriting manually defined version constraints.

Supersedes #20.

nm2107 commented on pkg/defkinds/php/definition.go:
```diff
@@ -595,83 +578,68 @@ func addIntegrations(stageDef *StageDefinition) error {
 //
 // This list has been obtained using:
 // docker run --rm -t php:7.2-fpm-buster php -r 'var_dump(get_loaded_extensions());'
-var preinstalledExtensions = []string{
```
what about also update the above comment to indicate what the struct is supposed to contain ?

NiR- replied:
Mh I can but using a map with empty struct/interface as value is quite a usual pattern in golang whenever a slice starts to be too big :wink:

nm2107 commented on pkg/defkinds/php/definition.go:
```diff
@@ -595,83 +578,68 @@ func addIntegrations(stageDef *StageDefinition) error {
-			def.Extensions["sockets"] = "*"
-		}
+	if def.Extensions.Has("soap") {
+		def.Extensions.Add("sockets", "*", false)
```
:+1:

nm2107 commented on pkg/builddef/builddef.go:
```diff
@@ -16,3 +16,78 @@ type BuildDef struct {
+// without overwriting any preexisting value. This is used in specialized
+// defkind handlers to add infered items without rewriting manually-defined
+// values.
+type VersionMap map[string]string
```
what about writing unit tests for this new struct and its impl ?

NiR- commented:
I added a commit which is introducing a new method named `Overwrite()`. It replaces the `overwrite` bool at the end of `Add()` method.

I also added proper tests for Add, Overwrite and Remove methods but not for others. There're already a lot of tests in the app (~7k loc of tests for ~5k of real code) and the others methods are called from defkind handlers so if they break, other parts of the code should too.
